cirq/ops/linear_combinations.py

```diff
@@ -205,3 +209,132 @@ def extend(expansion: value.LinearDict[str],
+
+    return True
+
+
+def _pauli_string_from_unit(unit: UnitPauliStringT,
+                            coefficient: Union[int, float, complex] = 1):
+    return PauliString(dict(unit), coefficient=coefficient)
+
+
+@value.value_equality(approximate=True)
+class PauliSum:
+    """Represents operator defined by linear combination of PauliStrings.
+
+    Since PauliStrings store their own coefficients, this class
+    does not implement the LinearDict interface. Instead, you can
+    add and subtract terms and then iterate over the resulting
+    (simplified) expression.
+
+    Under the hood, this class is backed by a LinearDict with coefficient-less
+    PauliStrings as keys. PauliStrings are reconstructed on-the-fly during
+    iteration.
+    """
+
+    def __init__(self, linear_dict: value.LinearDict[UnitPauliStringT]):
+        if not _is_linear_dict_of_unit_pauli_string(linear_dict):
+            raise ValueError(
+                "PauliSum constructor takes a LinearDict[UnitPauliStringT]. "
+                "Consider using PauliSum.from_pauli_strings() or adding and "
+                "subtracting PauliStrings")
+        self._linear_dict = linear_dict
+
+    def _value_equality_values_(self):
+        return self._linear_dict
+
+    @classmethod
+    def from_pauli_strings(cls, terms: Union[PauliString, List[PauliString]]) \
+            -> 'PauliSum':
+        if isinstance(terms, PauliString):
+            terms = [terms]
+        termdict = defaultdict(lambda: 0)
+        for pstring in terms:
+            key = frozenset(pstring._qubit_pauli_map.items())
+            termdict[key] += pstring.coefficient
+        return cls(linear_dict=value.LinearDict(termdict))
+
+    def copy(self):
+        factory = type(self)
+        return factory(self._linear_dict.copy())
+
+    def __iter__(self):
+        for vec, coeff in self._linear_dict.items():
+            yield _pauli_string_from_unit(vec, coeff)
+
+    def __len__(self) -> int:
+        return len(self._linear_dict)
+
+    def __iadd__(self, other):
+        if isinstance(other, (float, int, complex)):
+            other = PauliSum.from_pauli_strings(
+                [PauliString(coefficient=other)])
+        elif isinstance(other, PauliString):
+            other = PauliSum.from_pauli_strings([other])
+
+        self._linear_dict += other._linear_dict
+        return self
+
+    def __add__(self, other):
+        result = self.copy()
+        result += other
+        return result
+
+    def __isub__(self, other):
+        if isinstance(other, PauliString):
+            other = PauliSum.from_pauli_strings([other])
+        self._linear_dict -= other._linear_dict
+        return self
+
+    def __sub__(self, other):
+        result = self.copy()
+        result -= other
+        return result
+
+    def __neg__(self):
+        factory = type(self)
+        return factory(-self._linear_dict)
+
+    def __imul__(self, a: value.Scalar):
+        self._linear_dict *= a
```

Comment:
Should return `NotImplemented` for unrecognized types, and this should be tested.

Note that this means the type annotation is not correct. You can use `@overload` from typing to give the specific cases if you're feeling ambitious.

Reply:
why is the type annotation not correct?

Reply:
should I add the check to LinearDict, which this method dispatches to?

Reply:
The type annotation is technically not correct because the method has a defined result when `a` is not a Scalar. Or... well, I guess I don't know the exact semantics assumed by mypy. Maybe it considereds returning NotImplemented to be an error where the type annotations ar enot supposed to apply.

Reply:
returning `NotImplemented` is a way for python itself to throw a `TypeError` after exhausting all possible options

Reply:
If mypy is happy, I'm happy.